File: Inicio/Concorrente/C/goRoutine.c
#include <stdio.h>
#include <unistd.h> // Para a função sleep

// Função para simular a operação demorada na goroutine
void minhaGoroutine() {
    for (int i = 0; i < 5; i++) {
        printf("Goroutine executada: %d\n", i);
        usleep(500000); // Simula operação demorada em microssegundos
    }
}

int main() {
    // Inicia a goroutine
    pid_t pid = fork(); // Cria um novo processo

    if (pid == 0) { // Processo filho
        minhaGoroutine();
    } else if (pid > 0) { // Processo pai
        // Executa tarefa na função principal
        for (int i = 0; i < 5; i++) {
            printf("Funcao main executada: %d\n", i);
            usleep(200000); // Simula operação demorada em microssegundos
        }
    } else { // Se houve erro ao criar o processo filho
        perror("Erro ao criar processo filho");
        return 1;
    }

    sleep(1); // Aguarda um segundo para garantir que todas as operações terminem
    return 0;
}

File: Inicio/Concorrente/goRoutine.go
/*
Abstração:
Tipo: Comportamental.
Descrição: Abstração de comportamento ao ocultar os detalhes da execução assíncrona (goroutine) e da simulação de operações demoradas.
Encapsulamento:
Tipo: Operacional.
Descrição: Encapsulamento do tempo de espera necessário para sincronizar a execução das goroutines e da função main().
*/

package main

import (
    "fmt"
    "time"
)

func minhaGoroutine(){

    for i := 0; i < 5; i++{
        fmt.Println("Goroutine executada: ", i) // Abstração: Mensagens de impressão para fornecer feedback sobre a execução de uma goroutine.
        time.Sleep(time.Millisecond * 500) // Abstração: Simulação de uma operação demorada.
    }
}//end func

func main(){

    // Inicia a goroutine
    go minhaGoroutine() // Abstração: Inicia uma nova goroutine para execução assíncrona.

    // Executa tarefa na função principal

    for i := 0; i < 5; i++ {
        fmt.Println("Funcao main executada", i) // Abstração: Mensagens de impressão para fornecer feedback sobre a execução da função principal.
        time.Sleep(time.Millisecond * 200) // Abstração: Simulação de uma operação demorada.
    }

    time.Sleep(time.Second) // Encapsulamento: Garante que a execução não termine até que as goroutines tenham tido tempo suficiente para serem concluídas.
}//end main
